### ISSUE TITLE
Arm/Arm64 Docker Build

### ISSUE THREAD

**Issue by TheFlatworm** [airsonic-advanced/airsonic-advanced#359]
Would it be possible to add an Arm/Arm64 docker build?

**Comment by randomnicode** [comment]
Hey @TheFlatworm , want to submit some PRs for it? We're happy to add them

**Comment by TheFlatworm** [comment]
I'm a noob but looking into how to do this!

**Comment by jokeofweek** [comment]
I believe this should be fairly straightforward... would likely need a second Dockerfile (Dockerfile.armhf) and it should pull from a different root image.

Instead of

```
FROM adoptopenjdk:14-jre-hotspot
```

It should be

```
FROM arm32v7/adoptopenjdk:14-jre-hotspot
```

**Comment by yllekz** [comment]
Hi folks, does this exist? I'd like to deploy this to my Pi but I can't find an arm tag.

**Comment by randomnicode** [comment]
Since tini has now been taken out from the Dockerfile, you could build on your machine with buildx for crossplatform support.

https://github.com/airsonic-advanced/airsonic-advanced/pull/438 might provide a hint on how to do it

If you have support to build linux/arm64 (qemu and buildx):
-build airsonic-advanced (or put the war in install/docker/target/dependency as airsonic-main.war)
- arm64 run: docker buildx --platform linux/arm64 --load -f install/docker/Dockerfile -t airsonicadvanced/airsonic-advanced:arm64-latest install/docker
- armv7 run: docker buildx --platform linux/arm/v7 --load -f install/docker/Dockerfile -t airsonicadvanced/airsonic-advanced:armv7-latest install/docker

**Comment by nesasio93** [comment]
Building the image is something not everyone will be capable to do, especially when buildx is a experimental feature of docker. Are there plans to officially support arm/arm64? I would love to try this out on my RPI4! 😄

**Comment by yllekz** [comment]
Same honestly. I'm not a developer and I can't just "build it on my own machine" honestly.

**Comment by randomnicode** [comment]
> Building the image is something not everyone will be capable to do, especially when buildx is a experimental feature of docker.

> Same honestly. I'm not a developer and I can't just "build it on my own machine" honestly.

Fair enough.

okay, in the interest of expediting this, we can roll out an arm build, however it won't have gone through the same testing process. The war itself will be tested as part of the ci/cd, but the container image (for arm builds) may not be unless someone wants to contribute PRs

**Comment by nesasio93** [comment]
> okay, in the interest of expediting this, we can roll out an arm build, however it won't have gone through the same testing process. The war itself will be tested as part of the ci/cd, but the container image (for arm builds) may not be unless someone wants to contribute PRs

That's great news! Thank you!




> I believe this should be fairly straightforward... would likely need a second Dockerfile (Dockerfile.armhf) and it should pull from a different root image.
> 
> Instead of
> 
> ```
> FROM adoptopenjdk:14-jre-hotspot
> ```
> 
> It should be
> 
> ```
> FROM arm32v7/adoptopenjdk:14-jre-hotspot
> ```

If that is true the only difference should be the base image, so I would be confident in the arm version working as good as the amd64 version 😄

**Comment by randomnicode** [comment]
No, creating the Dockerfile isn't an issue (and the suggestion is being superseded by a multiplatform build anyway)

The bigger issue is testing of the container.

The base image isn't the only thing that is different, on each version we install the ffmpeg binaries etc (which is different because of the different base images)

**Comment by randomnicode** [comment]
As #463 was merged, airsonic-advanced should now be releasing multiplatform images to Docker hub now. You should be able to pull down the right architecture based on your docker platform

![image](https://user-images.githubusercontent.com/45723468/106636047-5eea8800-6536-11eb-9a6d-366c1dc51d36.png)

**Comment by randomnicode** [comment]
Please test and if it seems stable, we can close this issue.

**Comment by nesasio93** [comment]
It does not seem stable for me. If I access airsonic directly with IP:4040 on my Pi everything seems normal, but I can't play music. When I click "Play All" for an album all I see is this in my console:

![image](https://user-images.githubusercontent.com/66573084/106642530-bfa2b080-6588-11eb-918d-ae6e6631a6d3.png)

Play queue stays empty.

**Comment by nesasio93** [comment]
It's another story when I access airsonic over my reverse proxy. Music is playing, but the play queue remains empty. Console is full of errors like this:

![image](https://user-images.githubusercontent.com/66573084/106651354-56c13580-6594-11eb-8179-55ec5cca8e44.png)

There seems to be a problem with the websocket, but it says it has opend a websocket?!
![image](https://user-images.githubusercontent.com/66573084/106651497-82442000-6594-11eb-8798-ecf950700796.png)

**Comment by randomnicode** [comment]
> formatDuration is not defined

This does not seem like a docker issue but a browser issue, and no idea why it's happening with you especially since the function is defined. Perhaps you're blocking loading javascripts? if you look at the page structure, do you spot a `utils.js` import?

> problem with the websocket

Not sure which proxy you're using, but you need to set up your proxy rules right to allow sockjs through

**Comment by TheFlatworm** [comment]
I am getting this error: 

`OpenJDK Server VM warning: No monotonic clock was available - timed services may be adversely affected if the time-of-day clock changes`

**Comment by randomnicode** [comment]
> OpenJDK Server VM warning: No monotonic clock was available

A quick Google shows this is a problem with Ubuntu 20.04. You need to update libseccomp to 2.4.2 or later on your Raspberry Pi *host*. Latest version of libseccomp might not be available from the repo for Pi, but you need to find an alternative place to download and update it:

https://github.com/tianon/docker-brew-ubuntu-core/issues/196

https://github.com/AdoptOpenJDK/openjdk-docker/issues/469

https://github.com/AdoptOpenJDK/openjdk-docker/issues/479

**Comment by nesasio93** [comment]
> formatDuration is not defined

I started the server again and that error disappeared! Now everything works as expected. Weird, but nice! 😄 

> problem with the websocket

As you can see in the screenshot the error is shown, but the websocket is open and works. I'm using Traefik as a reverse proxy, because I like its simplicity and easy configuration with docker labels. Treafik supports websockets without any configuration. Only `JAVA_OPTS=-Dserver.forward-headers-strategy=native` is needed for airsonic-advanced to function correctly, it seems.

Looks like I'm switching from airsonic to airsonic-advanced 😎 Big thanks to all its developers!!!

**Comment by TheFlatworm** [comment]
>  You need to update libseccomp to 2.4.2 or later on your Raspberry Pi _host_.

I am running Debian with OMV and libseccomp is not installed at all.  I will install this and try again.

**Comment by TheFlatworm** [comment]
> A quick Google shows this is a problem with Ubuntu 20.04

Thank you - what was throwing me off was that I am running Debian buster which I now know has the same issue with an outdated libseccomp and that the correct package name is libseccomp2. After installing with dpkg everything seems to work!  I updated  JAVA_OPTS=-Xmx to about 3mb and I will run as a test for few weeks to test stability.  Thank you so much!

**Comment by qwertyAcabtino** [comment]
Tested with following setup & results:

# Setup

Rpi 2b. Docker container and version `airsonicadvanced/airsonic-advanced:edge-11.0.0-SNAPSHOT.20210217223709`

# Devices

When checking in the browser I got the same result as https://github.com/airsonic-advanced/airsonic-advanced/issues/359#issuecomment-771849585
When checking via music stash (the android app) the music plays perfectly.

**Comment by yllekz** [comment]
Reporting in on my Pi 4B with the "latest" tag. The app deploys just fine and I can configure it, but I cannot play music in the browser in Firefox on Windows. I was able to get music to play in Safari on iOS though.

**Comment by sunkup** [comment]
> Reporting in on my Pi 4B with the "latest" tag. The app deploys just fine and I can configure it, but I cannot play music in the browser in Firefox on Windows.
> 
> Edge on Windows can play just fine so I guess it's a Firefox problem. I was able to get music to play in Safari on iOS too.
> 
> Update: It might be a false alarm with either an extension or my Firefox Profile, as it works in a Private Window. I'll investigate this on my own.

got the same problem. Did you find out why? :)
